src/main/cpp/src/hive_hash.cu

```diff
@@ -210,11 +208,222 @@ class hive_device_row_hasher {
+     *
+     * And sublist structure factors in the hash value. For example, [[1, 0], [2]] and [[1], [0, 2]]
+     * have different hive hash values.
+     *
+     * The computation process for lists columns in this solution is as follows:
+     *            L1              List<list<int>>
+     *            |
+     *            L2              List<int>
+     *      /     |     \
+     *    L2[0] L2[1] L2[2]       List<int>
+     *     |            |
+     *    i1           i2         Int
+     *
+     * Note: L2、i1、i2 are all temporary columns, which would not be pushed into the stack.
+     * There is an optimization for the list column. If the child column is of primitive type, the
+     * hash value of the list column can be directly computed. Thus we can decrease the depth of
+     * the stack by one.
+     *
+     * @tparam T The type of the column.
+     * @param col The column device view.
+     * @param row_index The index of the row to compute the hash for.
+     * @return The computed hive hash value.
+     *
+     * @note This function is only enabled for nested column types.
+     */
     template <typename T, CUDF_ENABLE_IF(cudf::is_nested<T>())>
     __device__ hive_hash_value_t operator()(cudf::column_device_view const& col,
                                             cudf::size_type row_index) const noexcept
     {
-      CUDF_UNREACHABLE("Nested type is not supported");
+      cudf::column_device_view curr_col = col.slice(row_index, 1);
+      // The default constructor of `col_stack_frame` is deleted, so it can not allocate an array
+      // of `col_stack_frame` directly.
+      // Instead leverage the byte array to create the col_stack_frame array.
+      alignas(col_stack_frame) char stack_wrapper[sizeof(col_stack_frame) * MAX_NESTED_DEPTH];
+      col_stack_frame_ptr col_stack = reinterpret_cast<col_stack_frame_ptr>(stack_wrapper);
+      int stack_size                = 0;
+
+      col_stack[stack_size++] = col_stack_frame(curr_col);
+
+      while (stack_size > 0) {
+        col_stack_frame& top = col_stack[stack_size - 1];
+        curr_col             = top.get_element();
+        // Do not pop it until it is processed. The definition of `processed` is:
+        // - For structs, it is when all child columns are processed.
+        // - For lists, it is when all elements in the list are processed.
+        // - For primitive types, it is when the hash value is computed.
+        if (curr_col.type().id() == cudf::type_id::STRUCT) {
+          if (top.get_idx_to_process() == curr_col.num_child_columns()) {
+            if (--stack_size > 0) { col_stack[stack_size - 1].update_cur_hash(top.get_hash()); }
+          } else {
+            // Push the next child into the stack
+            col_stack[stack_size++] =
+              col_stack_frame(cudf::detail::structs_column_device_view(curr_col).get_sliced_child(
+                top.get_and_inc_idx_to_process()));
```

Comment:
This can be optimized further, in a similar way as lists column: 
```
while(child is primitive type) {
 // update hash directly without pushing child onto stack
 // move to next child
}
```